feat!: port to 0.2.0

arminius-smh opened the pull request:
Hello, I tried myself at porting this theme to the four color palettes, trying to solve https://github.com/catppuccin/thelounge/issues/1

I thought whisker might be nice to generate the package.json as well for each individual package soo the only thing that has to be done to upload them is run `just build` and then `npm publish` in each directory. (also tested to publish them myself and install them with `thelounge install` and it worked just fine, packages seem just to need the thelounge metadata tag and that's it to install them with that wrapper)
If this approach however is not desired, I can also try to port it to another build system.

Since I suppose this would end up to four different npm packages, one for each flavor, I used `just` to copy the readme, license and assets directory to each flavor again.
This technically isn't needed for anything, only it would also show the readme etc. on every package on npmjs.com, I guess that's just a matter of what you prefer, I just thought it would look nicer on npmjs.com.

Otherwise I didn't really change anything about the existing theme, I changed it to a mauve accent from the default the Lounge green for the buttons, sidepanel and links and also implemented the fix from https://github.com/catppuccin/thelounge/pull/2.

Let me know if there is anything with the theme itself or the build steps that needs to be fixed or is not desired!

uncenter commented on assets/preview.webp:
Do you have the original 4 previews used to create this? In the template we display all of them, like so: https://github.com/catppuccin/template#previews. I would add the original per-flavor previews back to this folder and copy paste that preview section from the file linked above.

arminius-smh replied:
Ah right sorry, I should have looked at an up to date version of the readme and not just update it here, will do that

uncenter commented on templates/theme.tera:
```diff
@@ -0,0 +1,410 @@
+    --window-bg-color: #{{base.hex}};
+    --body-color: #{{text.hex}};
+    --body-color-muted: #{{overlay1.hex}};
+    --link-color: #{{mauve.hex}};
```
To match the style guide at the moment I think these should be blue, cc @sgoudham?
```suggestion
    --link-color: #{{blue.hex}};
```

sgoudham replied:
I'd appreciate it if it was blue, yeah

arminius-smh replied:
okay, saw some discussion on discord about it but wasn't quite sure what is now desired, will do that together with new previews

uncenter commented:
No need to force push FYI, we usually squash merge so only the title really matters. Besides, force pushing makes it just a little more complicated to review!

uncenter commented on README.md:
```diff
@@ -12,13 +12,44 @@
-2. Select theme from the settings menu of your instance of The Lounge
+<details>
+<summary>🌻 Latte</summary>
+<img src="assets/latte.png"/>
+</details>
+<details>
+<summary>🪴 Frappé</summary>
+<img src="assets/frappe.png"/>
+</details>
+<details>
+<summary>🌺 Macchiato</summary>
+<img src="assets/macchiato.png"/>
+</details>
+<details>
+<summary>🌿 Mocha</summary>
+<img src="assets/mocha.png"/>
+</details>
```
Let's make them all webp!

arminius-smh commented:
> No need to force push FYI, we usually squash merge so only the title really matters. Besides, force pushing makes it just a little more complicated to review!

Oops, sorry makes sense, will do that from now on, I'm mostly used to just force push everything from working on personal projects so I didn't think about that

sgoudham commented:
Thanks for taking on this work! I believe every single port will be 0.2.0 after this :D 

I haven't had a chance to review the files yet but one concern for me is if there is a naming restriction on this package due to how `thelounge` queries NPM to find themes.

What option from the list below can we do? I've ordered them in ascending order of preference.

1. (most preferred) publish as `@catppuccin/thelounge` and provide all themes through this package
2. publish as `@catppuccin/thelounge-<flavour>` for each flavour of the theme.
3. publish as `@catppuccin/thelounge-theme-<flavour>` for each flavour of the theme.
4. (least preferred) publish as `thelounge-theme-catppuccin<flavour>` for each flavour of the theme. Unfortunately, this seems to be the (only?) option according to the [documentation](https://thelounge.chat/docs/guides/theme-creation) 

Can we get away with just doing `@catppuccin/thelounge` and distributing 4 themes under there? Or do we need to have a package for each variant of the theme?

arminius-smh commented:
> What option from the list below can we do? I've ordered them in ascending order of preference.

The general naming of the package itself doesn't matter, however it seems like their wrapper can't parse the scope in npm packages?
I published [@arminius-smh/thelounge-test](https://www.npmjs.com/package/@arminius-smh/thelounge-test)
However `thelounge install @arminius-smh/thelounge-test` or `thelounge install "@arminius-smh/thelounge-test"`

both lead to `2024-09-03 11:58:41 [ERROR] TypeError: Cannot read properties of undefined (reading 'arminius-smh/thelounge-test')`

(I also just saw that the npm packge resolves the images in the readme relative to the position from the specified git repo so I don't think the assets directory has to be replicated, so that's nice)

I only have to look into if it's possible to do `thelounge-theme-catppuccin` so that all themes would be under one package, if that would be a bit better than the 4th option?

uncenter commented:
Found https://github.com/thelounge/thelounge/issues/4295 in regards to the npm specifier/name parsing issues. No activity on it :/

arminius-smh commented:
Hmm I looked a bit into their code and the install error was rather easy to fix, they just used the split at '@' to get the version and never thought of scoped packages, with something like this: (everything else, uninstall and upgrade etc. work just fine because they just then read their package.json)
```diff
diff --git a/server/command-line/install.ts b/server/command-line/install.ts
index adce9618..7d656761 100644
--- a/server/command-line/install.ts
+++ b/server/command-line/install.ts
@@ -47,9 +47,21 @@ program
 				.readFile(path.join(packageName.substring("file:".length), "package.json"), "utf-8")
 				.then((data) => JSON.parse(data) as typeof packageJson);
 		} else {
-			const split = packageName.split("@");
-			packageName = split[0];
-			const packageVersion = split[1] || "latest";
+			let packageVersion = "latest"
+
+			if (packageName.startsWith("@")) {
+				const atIndex = packageName.lastIndexOf("@");
+
+				if (atIndex > 0) {
+					packageVersion = packageName.slice(atIndex + 1) || "latest";
+					packageName = packageName.slice(0, atIndex);
+				}
+			} else {
+				const split = packageName.split("@");
+				packageName = split[0];
+				packageVersion = split[1] || "latest";
+			}
 
 			readFile = packageJson.default(packageName, {
 				fullMetadata: true,
```
The second issue is that the theme doesn't get loaded, even if it is installed at the correct place, I haven't quite figured out where they parse the css, but this whole issue might be fixable upstream?

In the meantime this pr could either be blocked until this is fixed, or the `thelounge install` part and npm package could be left out for now, and later added when this is fixed, because it might take some time to get the pr into master and then wait for a new release, then at least option 2 - publish as `@catppuccin/thelounge-<flavour>` for each flavour of the theme would be feasible

arminius-smh commented:
Okay, made a pr upstream that should fix the issue: https://github.com/thelounge/thelounge/pull/4907

sgoudham commented:
> Soo to sum it up, @catppuccin/thelounge providing multiple themes won't work, it is one package per theme, although @catppuccin/thelounge-<flavour> will work if my pr gets merged

I'd be happy with this since it does come under our `@catppuccin` scope. Hope the upstream PR lands!

sgoudham commented:
I've just had a read of the documentation which says the following:

> Themes for The Lounge are CSS files bundled in packages preferably hosted on [the npm registry](https://www.npmjs.com/search?q=keywords:thelounge-theme).
**Alternatively, packages from the local file system can be installed.**

I'm happy to approve and merge this PR if the README is updated on how the theme can be installed from the local file system. We can track the NPM upload separately under #4

sgoudham commented on README.md:
```diff
@@ -12,13 +12,44 @@
-	<img src="https://raw.githubusercontent.com/catppuccin/thelounge/main/assets/demo.png"/>
+	<img src="assets/preview.webp"/>
 </p>
 
-## Usage
+## Previews
 
-1. Run `thelounge install thelounge-theme-catppuccin-mocha`
-2. Select theme from the settings menu of your instance of The Lounge
+<details>
+<summary>🌻 Latte</summary>
+<img src="assets/latte.webp"/>
+</details>
+<details>
+<summary>🪴 Frappé</summary>
+<img src="assets/frappe.webp"/>
+</details>
+<details>
+<summary>🌺 Macchiato</summary>
+<img src="assets/macchiato.webp"/>
+</details>
+<details>
+<summary>🌿 Mocha</summary>
+<img src="assets/mocha.webp"/>
+</details>
```
Since we're duplicating this README.md file into subdirectories, the relative links here will break so it's probably better to use absolute links (even if it's against our recommendations)

sgoudham commented on README.md:
````diff
@@ -12,13 +12,44 @@
+1. Copy the contents of `theme.css` from your desired flavor in `themes/catppuccin-{flavor}/`
+2. Paste it into your instance of `The Lounge` in Settings -> Appearance -> Custom Stylesheet
+
+### thelounge install
+
+1. Run `thelounge install @catppuccin/thelounge-theme-catppuccin-{flavor}` e.g.
+
+```bash
+thelounge install @catppuccin/thelounge-theme-catppuccin-mocha
+```
+
+2. Select the theme from your instance of `The Lounge` in Settings -> Appearance -> Theme
````
We can remove this for now and add it back in another PR under #4 
```suggestion
```

sgoudham commented on assets/preview-old.webp:
Not needed anymore

sgoudham reviewed (changes requested):
A few comments from me below.

Also, I noticed that the sidebar wasn't `mantle` in the screenshots:

![image](https://github.com/user-attachments/assets/4074a922-d3d1-462d-9f6b-1f88cd8734a3)

The relevant CSS should be updated to make the sidebar `mantle`.

sgoudham reviewed (changes requested):
One more thing I noticed in the screenshots:

The text colour on latte shouldn't change when it's selected, it should still be `text`:

![image](https://github.com/user-attachments/assets/fd3b8c26-8e0d-432c-a7ad-8a95f4afb067)

the selected background should also match our style guide, if possible:

![image](https://github.com/user-attachments/assets/fef6eddd-3086-4f43-aada-3dfdb4525619)

arminius-smh commented:
> The text colour on latte shouldn't change when it's selected, it should still be `text`:

Right, this was unstyled

There is probably more unstyled stuff that doesn't work well on the light theme or stuff that has to be changed to conform with the styleguide, as I said, I didn't really change anything from the og theme and mostly just put it in a tera file.

This then probably needs to be looked over more closely, I'll change it to a draft for now, I don't really have the time right now to do this properly.

sgoudham commented:
Funnily enough, I was just about to approve it since I was just going off the screenshots. 

We can merge this one and raise future PRs whenever you or someone else has more time to go through the application and see what may need themed.

Thank you so much for your work!

arminius-smh commented:
Okay, sure!
